[Long-format table annotation Part 2] annotator R API

logstar opened the pull request:
### Purpose/implementation Section

#### What scientific question is your analysis addressing?

Create application programming interface (API) and command line interface (CLI) for handling [long-format tables](https://en.wikipedia.org/wiki/Wide_and_narrow_data#Narrow) that are generated by analysis modules. API provides analysis module developers with functions that can be imported into their own scripts via R `source('path/to/the/function/file.R')` or Python `import os, sys; sys.path.append(os.path.abspath("path/to/the/function/dir")); import function_filename_but_no_dot_py`. CLI provides analysis module developers with scripts that can be executed in their own run-module shell script with either `Rscript --vanilla path/to/the/script.R arg long.tsv long_edited.tsv` or `python3 path/to/the/script.py arg long.tsv long_edited.tsv`.

This module is suggested by @jharenza and @kgaonkar6 in Slack at <https://opentargetspediatrics.slack.com/archives/C021Z53SK98/p1626290031138100?thread_ts=1626287625.133600&cid=C021Z53SK98>, in order to alleviate the burdens of analysis module developers for adding annotations and keeping track of what annotations need to be added. This module could also potentially handle large file storage issues at a later point, since the file size limit of GitHub is 100MB.

| Sub-module name                                               | Implemented function                    | Available interface(s)                                                                                 |
|---------------------------------------------------------------|-----------------------------------------|--------------------------------------------------------------------------------------------------------|
| `annotator` | Add gene and `cancer_group` annotations | R API, R CLI WIP |

(Update Jul 20 12:16 2021 YZ)

This pull request is the second part of the `long-format-table-utils` module, which implements the `annotator` R API.

#### What was your approach?

The `long-format-table-utils/annotator/annotator-api.R` file provides the `annotate_long_format_table` function for annotating long-format tables with the following steps.

Check input long-format tables have the following columns:

| Column name       | Description                                                                                                                                              |
|-------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------|
| `Gene_symbol`     | HUGO symbols, e.g. PHLPP1, TM6SF1, and DNAH5.                                                                                                            |
| `Gene_Ensembl_ID` | Ensembl ENSG IDs without `.#` versions, e.g. ENSG00000039139, ENSG00000111261, and ENSG00000169710                                                       |
| `Disease`         | The `cancer_group` in the `histologies.tsv`, e.g. Adamantinomatous Craniopharyngioma, Atypical Teratoid Rhabdoid Tumor, and Low-grade glioma/astrocytoma |

Add one or more of the following gene annotations, by specifying the `columns_to_add` parameter in the `annotate_long_format_table` function.

(Update Jul 19 12:15 2021 YZ)

| Annotation column name | Non-missing value                                                                                                                                                                                                                                                      | Source data                                                                                         |
|------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------|
| `RMTL`                 | `Relevant Molecular Target (RMTL version 1)` or `Non-Relevant Molecular Target (RMTL version 1)`                                                                                                                                                                       | `data/ensg-hugo-rmtl-v1-mapping.tsv`                                                                |
| `Gene_type`            | A sorted comma separated list of one or more of the following gene types: `CosmicCensus`, `Kinase`, `Oncogene`, `TranscriptionFactor`, and `TumorSuppressorGene`. Example values: `CosmicCensus`, `CosmicCensus,Kinase`, and `CosmicCensus,Kinase,TumorSuppressorGene`. | `analyses/fusion_filtering/references/genelistreference.txt`                                        |
| `OncoKB_cancer_gene`   | `Y` or `N`                                                                                                                                                                                                                                                             | `analyses/long-format-table-utils/annotator/annotation-data/oncokb-cancer-gene-list.tsv`            |
| `OncoKB_oncogene_TSG`  | `Oncogene`, or `TumorSuppressorGene`, or `Oncogene,TumorSuppressorGene`                                                                                                                                                                                                | `analyses/long-format-table-utils/annotator/annotation-data/oncokb-cancer-gene-list.tsv`            |
| `Gene_full_name`       | A single string of gene full name, e.g. `cytochrome c oxidase subunit III` and `ATP synthase F0 subunit 6`                                                                                                                                                             | `analyses/long-format-table-utils/annotator/annotation-data/ensg-gene-full-name-refseq-protein.tsv` |
| `Protein_RefSeq_ID`    | A sorted comma separated list of one or more protein RefSeq IDs, e.g. `NP_004065.1`, `NP_000053.2,NP_001027466.1`, and `NP_000985.1,NP_001007074.1,NP_001007075.1`                                                                                                     | `analyses/long-format-table-utils/annotator/annotation-data/ensg-gene-full-name-refseq-protein.tsv` |


The `RMTL` information is obtained from PediatricOpenTargets/OpenPedCan-analysis data release.

Note: only add `Gene_type` to gene-level tables, which can be implemented by leaving `"Gene_type"` out of the `columns_to_add` parameter of the `annotate_long_format_table` function in `annotator-api.R`. The `Gene_type` information is obtained from `../fusion_filtering/references/genelistreference.txt`, and its sources are described at <https://github.com/d3b-center/annoFuse#prerequisites-for-cohort-level-analysis>.

The `OncoKB_cancer_gene` and `OncoKB_oncogene_TSG` information is listed in `annotator/annotation-data/oncokb-cancer-gene-list.tsv`, which is downloaded from <https://www.oncokb.org/cancerGenes>. The last update of the table is on 06/16/2021. To update the table, re-download the updated table from <https://www.oncokb.org/cancerGenes>.

The `Gene_full_name` and `Protein_RefSeq_ID` information is downloaded from <https://mygene.info/> using the [mygene package](https://bioconductor.org/packages/release/bioc/html/mygene.html).

Add the following disease(/`cancer_group`) annotations:

(Update Jul 19 12:15 2021 YZ)

| Annotation column name | Non-missing value                                                                         | Source data              |
|------------------------|-------------------------------------------------------------------------------------------|--------------------------|
| `EFO`                  | A single string of EFO code, e.g. `EFO_1000069`, `EFO_1002008`, and `EFO_1000177`         | `data/efo-mondo-map.tsv` |
| `MONDO`                | A single string of MONDO code, e.g. `MONDO_0002787`, `MONDO_0020560`, and `MONDO_0009837` | `data/efo-mondo-map.tsv` |


The `EFO` and `MONDO` information is obtained from PediatricOpenTargets/OpenPedCan-analysis data release.

(Update Jul 19 11:13 2021 YZ) Use the long-format table annotator API in an analysis module with the following steps:

1. Change the working directory of the analysis module to be `OpenPedCan-analysis` or a subdirectory of `OpenPedCan-analysis`. This allows the API function `annotate_long_format_table` to locate annotation data files.
2. `source` the `long-format-table-utils/annotator/annotator-api.R` file.
3. If the class of the table to be annotated is not `tibble::tbl_df`, convert the table to `tibble::tbl_df` with `tibble::as_tibble`. After conversion, carefully check rownames, colnames, column classes (especially factors), and other properties that may affect the correctness of you code.
4. If `c("Gene_symbol", "Gene_Ensembl_ID", "Disease")` are not all present in the colnames of the table to be annotated, add new columns or rename existing ones to have all required columns.
5. Call `annotate_long_format_table` to add one or more of the available annotation columns, by specifying the `columns_to_add` parameter in the `annotate_long_format_table` function. Read the documentation comment of the function for usage.
6. Rename, select, and reorder the columns of the annotated table for output.

Following is an example usage in the `rna-seq-expression-summary-stats` module `01-tpm-summary-stats.R`.

```text
> getwd()
[1] "/home/rstudio/OpenPedCan-analysis/analyses/rna-seq-expression-summary-stats"
> source('../long-format-table-utils/annotator/annotator-api.R')
> class(m_tpm_ss_long_tbl)
[1] "tbl_df"     "tbl"        "data.frame"
> colnames(m_tpm_ss_long_tbl)
 [1] "gene_symbol"                          "gene_id"                             
 [3] "cancer_group"                         "cohort"                              
 [5] "tpm_mean"                             "tpm_sd"                              
 [7] "tpm_mean_cancer_group_wise_zscore"    "tpm_mean_gene_wise_zscore"           
 [9] "tpm_mean_cancer_group_wise_quantiles" "n_samples"                           
> ann_m_tpm_ss_long_tbl <- m_tpm_ss_long_tbl %>%
+   rename(Gene_symbol = gene_symbol, Gene_Ensembl_ID = gene_id,
+          Disease = cancer_group)
> ann_m_tpm_ss_long_tbl <- annotate_long_format_table(
+   ann_m_tpm_ss_long_tbl, columns_to_add = c('MONDO', 'RMTL', 'EFO'))
> m_tpm_ss_long_tbl <- ann_m_tpm_ss_long_tbl %>%
+   rename(gene_symbol = Gene_symbol, gene_id = Gene_Ensembl_ID,
+          cancer_group = Disease) %>%
+   select(gene_symbol, RMTL, gene_id,
+          cancer_group, EFO, MONDO, n_samples, cohort,
+          tpm_mean, tpm_sd,
+          tpm_mean_cancer_group_wise_zscore, tpm_mean_gene_wise_zscore,
+          tpm_mean_cancer_group_wise_quantiles)
```

#### What GitHub issue does your pull request address?

<https://github.com/PediatricOpenTargets/ticket-tracker/issues/112>

### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

#### Which areas should receive a particularly close look?

Whether the code in `annotator/annotator-api.R` works as expected?

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes.

The R CLI part is still a WIP, so please disregard the relevant content in the README.md, as they may subject to major changers.

### Results

#### What types of results are included (e.g., table, figure)?

No result.

#### What is your summary of the results?

No result.

#### Reproducibility Checklist

<!-- Check all those that apply or remove this section if it is not applicable.-->

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [NA. CI is not available yet.] This analysis has been added to continuous integration.

#### Documentation Checklist

<!-- Please review https://github.com/AlexsLemonade/OpenPBTA-analysis#documenting-your-analysis -->

- [x] This analysis module has a `README` and it is up to date.
- [x] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

jharenza reviewed (commented):
Hi @logstar - was just going through this. I noticed you have `Gene_Ensembl_ID` as a required input column, but I think we will want to add this as an annotation based on the hugo symbol to minimize people needing the mapping file, correct?

logstar commented:
> Hi @logstar - was just going through this. I noticed you have `Gene_Ensembl_ID` as a required input column, but I think we will want to add this as an annotation based on the hugo symbol to minimize people needing the mapping file, correct?

This is correct. I required both `Gene_Ensembl_ID` and `Gene_symbol`, mainly because some `Gene_symbol`s are mapped to multiple `Gene_Ensembl_ID`s, and I think implicitly creating duplicates in the table may be unwanted by the users.

We could definitely require only `Gene_Ensembl_ID` or `Gene_symbol`.

jharenza reviewed (commented):
Overall, this works well - my other question is about input/output column names. In your example, you have renamed the columns for input and then renamed them back for output:

```
> getwd()
[1] "/home/rstudio/OpenPedCan-analysis/analyses/rna-seq-expression-summary-stats"
> source('../long-format-table-utils/annotator/annotator-api.R')
> class(m_tpm_ss_long_tbl)
[1] "tbl_df"     "tbl"        "data.frame"
> colnames(m_tpm_ss_long_tbl)
 [1] "gene_symbol"                          "gene_id"                             
 [3] "cancer_group"                         "cohort"                              
 [5] "tpm_mean"                             "tpm_sd"                              
 [7] "tpm_mean_cancer_group_wise_zscore"    "tpm_mean_gene_wise_zscore"           
 [9] "tpm_mean_cancer_group_wise_quantiles" "n_samples"                           
> ann_m_tpm_ss_long_tbl <- m_tpm_ss_long_tbl %>%
+   rename(Gene_symbol = gene_symbol, Gene_Ensembl_ID = gene_id,
+          Disease = cancer_group)
> ann_m_tpm_ss_long_tbl <- annotate_long_format_table(
+   ann_m_tpm_ss_long_tbl, columns_to_add = c('MONDO', 'RMTL', 'EFO'))
> m_tpm_ss_long_tbl <- ann_m_tpm_ss_long_tbl %>%
+   rename(gene_symbol = Gene_symbol, gene_id = Gene_Ensembl_ID,
+          cancer_group = Disease) %>%
+   select(gene_symbol, RMTL, gene_id,
+          cancer_group, EFO, MONDO, n_samples, cohort,
+          tpm_mean, tpm_sd,
+          tpm_mean_cancer_group_wise_zscore, tpm_mean_gene_wise_zscore,
+          tpm_mean_cancer_group_wise_quantiles)
```

It seems we should standardize the requirement so there is no renaming needed, i.e. - all modules will output the same column headers for this? 

I ran the below and it worked perfectly:

```
library(tidyverse)
zscores <- read_tsv("~/OT/analyses/rna-seq-expression-summary-stats/results/long_n_tpm_mean_sd_quantile_gene_wise_zscore.tsv.gz") %>%
   rename(Gene_symbol = gene_symbol, Gene_Ensembl_ID = gene_id, Disease = cancer_group) %>%
  select(Gene_symbol, Gene_Ensembl_ID, cohort, Disease, tpm_mean, tpm_sd, n_samples, tpm_mean_gene_wise_zscore, tpm_mean_cancer_group_wise_quantiles) 

class(zscores)
colnames(zscores)

ann_zscores <- annotate_long_format_table(zscores, columns_to_add = c('MONDO', 'RMTL', 'EFO'))
head(ann_zscores)
```

logstar commented:
> Overall, this works well - my other question is about input/output column names. In your example, you have renamed the columns for input and then renamed them back for output:
> 
> ```
> > getwd()
> [1] "/home/rstudio/OpenPedCan-analysis/analyses/rna-seq-expression-summary-stats"
> > source('../long-format-table-utils/annotator/annotator-api.R')
> > class(m_tpm_ss_long_tbl)
> [1] "tbl_df"     "tbl"        "data.frame"
> > colnames(m_tpm_ss_long_tbl)
>  [1] "gene_symbol"                          "gene_id"                             
>  [3] "cancer_group"                         "cohort"                              
>  [5] "tpm_mean"                             "tpm_sd"                              
>  [7] "tpm_mean_cancer_group_wise_zscore"    "tpm_mean_gene_wise_zscore"           
>  [9] "tpm_mean_cancer_group_wise_quantiles" "n_samples"                           
> > ann_m_tpm_ss_long_tbl <- m_tpm_ss_long_tbl %>%
> +   rename(Gene_symbol = gene_symbol, Gene_Ensembl_ID = gene_id,
> +          Disease = cancer_group)
> > ann_m_tpm_ss_long_tbl <- annotate_long_format_table(
> +   ann_m_tpm_ss_long_tbl, columns_to_add = c('MONDO', 'RMTL', 'EFO'))
> > m_tpm_ss_long_tbl <- ann_m_tpm_ss_long_tbl %>%
> +   rename(gene_symbol = Gene_symbol, gene_id = Gene_Ensembl_ID,
> +          cancer_group = Disease) %>%
> +   select(gene_symbol, RMTL, gene_id,
> +          cancer_group, EFO, MONDO, n_samples, cohort,
> +          tpm_mean, tpm_sd,
> +          tpm_mean_cancer_group_wise_zscore, tpm_mean_gene_wise_zscore,
> +          tpm_mean_cancer_group_wise_quantiles)
> ```
> 
> It seems we should standardize the requirement so there is no renaming needed, i.e. - all modules will output the same column headers for this?
> 
> I ran the below and it worked perfectly:
> 
> ```
> library(tidyverse)
> zscores <- read_tsv("~/OT/analyses/rna-seq-expression-summary-stats/results/long_n_tpm_mean_sd_quantile_gene_wise_zscore.tsv.gz") %>%
>    rename(Gene_symbol = gene_symbol, Gene_Ensembl_ID = gene_id, Disease = cancer_group) %>%
>   select(Gene_symbol, Gene_Ensembl_ID, cohort, Disease, tpm_mean, tpm_sd, n_samples, tpm_mean_gene_wise_zscore, tpm_mean_cancer_group_wise_quantiles) 
> 
> class(zscores)
> colnames(zscores)
> 
> ann_zscores <- annotate_long_format_table(zscores, columns_to_add = c('MONDO', 'RMTL', 'EFO'))
> head(ann_zscores)
> ```

Thank you for testing this PR @jharenza !

I agree that we should have standardized column names for all tables that are prepared for the FNL team. The example demonstrates a quick solution for module developers to handle the required columns.

jharenza commented:
> > Hi @logstar - was just going through this. I noticed you have `Gene_Ensembl_ID` as a required input column, but I think we will want to add this as an annotation based on the hugo symbol to minimize people needing the mapping file, correct?
> 
> This is correct. I required both `Gene_Ensembl_ID` and `Gene_symbol`, mainly because some `Gene_symbol`s are mapped to multiple `Gene_Ensembl_ID`s, and I think implicitly creating duplicates in the table may be unwanted by the users.
> 
> We could definitely require only `Gene_Ensembl_ID` or `Gene_symbol`.

Hmm, maybe I missed this in `rna-seq-expression-summary-stats` - how are you handling the duplicates? Do we need to have other modules do the same or simply map the ENSG ids?

NHJohnson commented on analyses/long-format-table-utils/annotator/annotator-api.R:
```diff
@@ -0,0 +1,358 @@
+          char_vec_to_str(available_ann_columns), "."))
+      }
+      if (col_add %in% colnames(long_format_table)) {
+        stop(paste0(col_add, " is already in the input long_format_table."))
```
Is it desirable to quit in this case? I can imagine just removing a column name from the columns to be added if it's already in the table, maybe with a warning. That way the user doesn't have to check if a column is already present, but maybe there is a reason for doing it the way you have it.

logstar replied:
As a user of this module as well, I would prefer the function to quit without touching the input table, if there is any column to be added already exist. Implicitly dropping a column and adding a new column may cause unexpected behaviors, even though not likely.

NHJohnson commented on analyses/long-format-table-utils/annotator/annotator-api.R:
```diff
@@ -0,0 +1,358 @@
+  # Keep the order of columns for the returned table
+  ret_tbl_col_order <- c(colnames(long_format_table), columns_to_add)
+
+  # Detect the ".git" folder -- this will in the project root directory. Use
```
will *be* in

logstar replied:
Thank you for pointing out the typo. Will fix soon.

NHJohnson commented on analyses/long-format-table-utils/annotator/annotator-api.R:
```diff
@@ -0,0 +1,358 @@
+    guess_max = 100000))
+
+
+  # Check no NA or duplicate in the key columns of annotation data tables
```
There is some code duplication here as multiple tables are checked for NAs and duplication. Would it make sense to create NA- and duplication-checking functions?

logstar replied:
I think creating functions won't help much, as the checkings are simple if statements.

The code duplication could be reduced by refactoring the whole R file by creating a list of all annotation tables, which may be worth doing. I will look into it.

logstar commented:
> > > Hi @logstar - was just going through this. I noticed you have `Gene_Ensembl_ID` as a required input column, but I think we will want to add this as an annotation based on the hugo symbol to minimize people needing the mapping file, correct?
> > 
> > 
> > This is correct. I required both `Gene_Ensembl_ID` and `Gene_symbol`, mainly because some `Gene_symbol`s are mapped to multiple `Gene_Ensembl_ID`s, and I think implicitly creating duplicates in the table may be unwanted by the users.
> > We could definitely require only `Gene_Ensembl_ID` or `Gene_symbol`.
> 
> Hmm, maybe I missed this in `rna-seq-expression-summary-stats` - how are you handling the duplicates? Do we need to have other modules do the same or simply map the ENSG ids?

I added duplicated rows for `Gene_symbol`s that are mapped to multiple `Gene_Ensembl_ID`s, and an example is noted at <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/27#issuecomment-868781566>, as I quoted below. This is implemented at line 289 of the code below.

For RNA modules that use collapsed TPM/count matrices with only `Gene_symbol`s available, adding duplicated rows as the `rna-seq-expression-summary-stats` may be applicable. For CNV and SNV modules, the consensus files have both `Gene_symbol`s and `Gene_Ensembl_ID`s.

<https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/7fb11a020a92d06c8685736546e860bfe23da7e2/analyses/rna-seq-expression-summary-stats/01-tpm-summary-stats.R#L286-L291>

> 
> Different from the wide tables, the `gene_id` column does not contain comma-separated ENSG IDs. If one gene symbol matches to multiple Ensembl IDs, each Ensembl gene ID will become one row in the long table. For example:
> 
> * Wide table:
> 
> ```
> # cancer_group_individual_cohort_mean_tpm.tsv
> gene_symbol    gene_id    Adamantimomatous craniopharyngioma___PBTA    Atypical meningioma___PBTA    Atypical Teratoid Rhabdoid Tumor___PBTA
> CDR1    ENSG00000184258.6,ENSG00000281508.1    111.1765    31.932000000000002    332.2896153846154
> ```
> 
> * Long table:
> 
> ```
> # long_n_tpm_mean_sd_quantile_group_wise_zscore.tsv.gz
> CDR1    ENSG00000184258.6    Adamantimomatous craniopharyngioma    PBTA    111.1765    181.70448082054318    0.11408800038186823    Highest expressed 25%    20
> CDR1    ENSG00000281508.1    Adamantimomatous craniopharyngioma    PBTA    111.1765    181.70448082054318    0.11408800038186823    Highest expressed 25%    20
> CDR1    ENSG00000184258.6    Atypical meningioma    PBTA    31.932000000000002    50.10366423725914    0.016467672549703015    Highest expressed 25%    5
> CDR1    ENSG00000281508.1    Atypical meningioma    PBTA    31.932000000000002    50.10366423725914    0.016467672549703015    Highest expressed 25%    5
> CDR1    ENSG00000184258.6    Atypical Teratoid Rhabdoid Tumor    PBTA    332.2896153846154    1075.378786603049    0.37106139758788453    Highest expressed 25%    26
> CDR1    ENSG00000281508.1    Atypical Teratoid Rhabdoid Tumor    PBTA    332.2896153846154    1075.378786603049    0.37106139758788453    Highest expressed 25%    26
> ```

NHJohnson commented on analyses/long-format-table-utils/annotator/annotator-api.R:
```diff
@@ -0,0 +1,358 @@
+  #
+  # pp in variable names means preprocessed
+  #
+  # %>% is not used, so the user's environment will not be modified by importing
```
This seems counterintuitive to me as `%>%` *is* used, a few lines down. Would you be able to clarify?

logstar replied:
Thank you for pointing this out. I forgot to remove this comment when refactoring.

NHJohnson commented on analyses/long-format-table-utils/annotator/annotator-api.R:
```diff
@@ -0,0 +1,358 @@
+    cgroup_efo_mondo_df,
+    Disease = cancer_group, EFO = efo_code, MONDO = mondo_code)
+
+  # # Code for checking data
```
Remove? Implement as tests?

logstar replied:
Will remove.

jharenza commented:
>Different from the wide tables, the gene_id column does not contain comma-separated ENSG IDs. If one gene symbol matches to multiple Ensembl IDs, each Ensembl gene ID will become one row in the long table. For example:

Ok, this makes sense since the backend of OT portal uses the ENSG id and those are unique.

logstar commented:
Thank you again for reviewing and test-running @jharenza !

I will add a note to the one symbol -> multiple ENSG ID mapping issue in the README.md and documentation comment of the API function. Then, I will merge this PR, so the SNV, TPM z-score, and fusion modules could be updated to use this API.

NHJohnson reviewed (commented):
I have made some comments. In following along your example, I found that at the step
```
> ann_m_tpm_ss_long_tbl <- annotate_long_format_table(
+     ann_m_tpm_ss_long_tbl, columns_to_add = c('MONDO', 'RMTL', 'EFO'))
```
there was
```
Error in .f(.x[[i]], ...) : 
  MONDO is already in the input long_format_table.
```

In addition to the provided instructions I had done
`source('01-tpm-summary-stats.R')`
to create `m_tpm_ss_long_tbl`

logstar commented:
> I have made some comments. In following along your example, I found that at the step
> 
> ```
> > ann_m_tpm_ss_long_tbl <- annotate_long_format_table(
> +     ann_m_tpm_ss_long_tbl, columns_to_add = c('MONDO', 'RMTL', 'EFO'))
> ```
> 
> there was
> 
> ```
> Error in .f(.x[[i]], ...) : 
>   MONDO is already in the input long_format_table.
> ```
> 
> In addition to the provided instructions I had done
> `source('01-tpm-summary-stats.R')`
> to create `m_tpm_ss_long_tbl`

Thank you for reviewing and test-running this PR @NHJohnson . 

The following part of the example indicates that the `MONDO` column is not in `m_tpm_ss_long_tbl`.

```text
> colnames(m_tpm_ss_long_tbl)
 [1] "gene_symbol"                          "gene_id"                             
 [3] "cancer_group"                         "cohort"                              
 [5] "tpm_mean"                             "tpm_sd"                              
 [7] "tpm_mean_cancer_group_wise_zscore"    "tpm_mean_gene_wise_zscore"           
 [9] "tpm_mean_cancer_group_wise_quantiles" "n_samples"          
```

I will add a note in the README.md that the columns to be added should not be in the input table.

logstar commented:
Added WIP label to this PR.

I will make some updates according to the reviews. I will also refactor the API to reduce code duplication.

logstar commented:
@jharenza @NHJohnson I have made some updates to this PR.

- Fix a bug in `annotate_long_format_table` that can add duplicated annotation columns. See commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/56/commits/81a041e7fff0a512053e0319b358c8e233fad8ca> for more details.
- Added tests to `annotate_long_format_table`. See commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/56/commits/7e38c190c7dce7612bf7f9abed4ad411c83bc31a> for more details.
- Refactored the `annotate_long_format_table` to reduce code duplication, as suggested by @NHJohnson, with functional programming paradigm. See commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/56/commits/3292b7f76015c7685057b996f77c3423bd6d6b7f> for more details.
- Added some notes, removed some comments, and fixed some typos according to the reviews.
- Enabled project root finding in git linked working trees. See commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/56/commits/fd99539295d51badc65c56afa751928edc2601c5> for more details.
- Removed `git diff --stat 'annotation-data'` command in `run-download-annotation-data.sh` to be more compatible with git linked work trees, see commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/56/commits/a85a97d71b91c54c398f8e6f5f709b7a485e71de> for more details.

NHJohnson commented:
Thanks for making these changes! I tested the functionality as before and it seemed ok, but I got some failures running the unit tests:
```
./run-tests.sh 
✔ |  OK F W S | Context
✖ |  26 5     | tests/test_annotate_long_format_table.R [4.3 s]
────────────────────────────────────────────────────────────────────────────────
test_annotate_long_format_table.R:25: failure: (unknown)
annotate_long_format_table(long_format_tibble) not equal to `inspected_annotated_long_format_tibble`.
Attributes: < Component “class”: Lengths (3, 4) differ (string compare on first 3) >
Attributes: < Component “class”: 3 string mismatches >

test_annotate_long_format_table.R:170: failure: (unknown)
annotate_long_format_table(...) not equal to `inspected_annotated_long_format_tibble`.
Names: 5 string mismatches
Attributes: < Component “class”: Lengths (3, 4) differ (string compare on first 3) >
Attributes: < Component “class”: 3 string mismatches >
Component 8: 6 string mismatches
Component 9: 6 string mismatches
Component 10: 6 string mismatches
Component 11: 6 string mismatches
Component 13: 6 string mismatches

test_annotate_long_format_table.R:177: failure: (unknown)
annotate_long_format_table(...) not equal to dplyr::select(inspected_annotated_long_format_tibble, -OncoKB_oncogene_TSG).
Names: 4 string mismatches
Component 8: 6 string mismatches
Component 9: 6 string mismatches
Component 10: 6 string mismatches
Component 12: 6 string mismatches

test_annotate_long_format_table.R:187: failure: (unknown)
annotate_long_format_table(...) not equal to dplyr::select(inspected_annotated_long_format_tibble, -RMTL).
Names: 6 string mismatches
Component 7: 6 string mismatches
Component 8: 6 string mismatches
Component 9: 6 string mismatches
Component 10: 6 string mismatches
Component 11: 6 string mismatches
Component 12: 6 string mismatches

test_annotate_long_format_table.R:197: failure: (unknown)
annotate_long_format_table(...) not equal to dplyr::select(inspected_annotated_long_format_tibble, -RMTL).
Names: 5 string mismatches
Component 8: 6 string mismatches
Component 9: 6 string mismatches
Component 10: 6 string mismatches
Component 11: 6 string mismatches
Component 12: 6 string mismatches
────────────────────────────────────────────────────────────────────────────────
✔ |   8       | tests/test_collapse_name_vec.R
✔ |   7       | tests/test_collapse_rp_lists.R
✔ |   5       | tests/test_collapse_rp_lists.R

══ Results ═════════════════════════════════════════════════════════════════════
Duration: 4.6 s

OK:       46
Failed:   5
Warnings: 0
Skipped:  0

Frustration is a natural part of programming :)
Done running run-tests.sh
```

Can you advise?

logstar commented:
> Thanks for making these changes! I tested the functionality as before and it seemed ok, but I got some failures running the unit tests:
> 
> ```
> ./run-tests.sh 
> ✔ |  OK F W S | Context
> ✖ |  26 5     | tests/test_annotate_long_format_table.R [4.3 s]
> ────────────────────────────────────────────────────────────────────────────────
> test_annotate_long_format_table.R:25: failure: (unknown)
> annotate_long_format_table(long_format_tibble) not equal to `inspected_annotated_long_format_tibble`.
> Attributes: < Component “class”: Lengths (3, 4) differ (string compare on first 3) >
> Attributes: < Component “class”: 3 string mismatches >
> 
> test_annotate_long_format_table.R:170: failure: (unknown)
> annotate_long_format_table(...) not equal to `inspected_annotated_long_format_tibble`.
> Names: 5 string mismatches
> Attributes: < Component “class”: Lengths (3, 4) differ (string compare on first 3) >
> Attributes: < Component “class”: 3 string mismatches >
> Component 8: 6 string mismatches
> Component 9: 6 string mismatches
> Component 10: 6 string mismatches
> Component 11: 6 string mismatches
> Component 13: 6 string mismatches
> 
> test_annotate_long_format_table.R:177: failure: (unknown)
> annotate_long_format_table(...) not equal to dplyr::select(inspected_annotated_long_format_tibble, -OncoKB_oncogene_TSG).
> Names: 4 string mismatches
> Component 8: 6 string mismatches
> Component 9: 6 string mismatches
> Component 10: 6 string mismatches
> Component 12: 6 string mismatches
> 
> test_annotate_long_format_table.R:187: failure: (unknown)
> annotate_long_format_table(...) not equal to dplyr::select(inspected_annotated_long_format_tibble, -RMTL).
> Names: 6 string mismatches
> Component 7: 6 string mismatches
> Component 8: 6 string mismatches
> Component 9: 6 string mismatches
> Component 10: 6 string mismatches
> Component 11: 6 string mismatches
> Component 12: 6 string mismatches
> 
> test_annotate_long_format_table.R:197: failure: (unknown)
> annotate_long_format_table(...) not equal to dplyr::select(inspected_annotated_long_format_tibble, -RMTL).
> Names: 5 string mismatches
> Component 8: 6 string mismatches
> Component 9: 6 string mismatches
> Component 10: 6 string mismatches
> Component 11: 6 string mismatches
> Component 12: 6 string mismatches
> ────────────────────────────────────────────────────────────────────────────────
> ✔ |   8       | tests/test_collapse_name_vec.R
> ✔ |   7       | tests/test_collapse_rp_lists.R
> ✔ |   5       | tests/test_collapse_rp_lists.R
> 
> ══ Results ═════════════════════════════════════════════════════════════════════
> Duration: 4.6 s
> 
> OK:       46
> Failed:   5
> Warnings: 0
> Skipped:  0
> 
> Frustration is a natural part of programming :)
> Done running run-tests.sh
> ```
> 
> Can you advise?

I wonder if you run the tests in the Docker image/container. All tests passed in the Docker image/container, so I guess it's R version and dependency issue. As all users of this module use the Docker image/container, it is acceptable if any test fails in other environment, but I will look further to see what exactly happened.

Run in the Docker image/container:

```text
$ bash analyses/long-format-table-utils/annotator/run-tests.sh✔ |  OK F W S | Context
✔ |  31       | tests/test_annotate_long_format_table.R [3.4 s]
✔ |   8       | tests/test_collapse_name_vec.R
✔ |   7       | tests/test_collapse_rp_lists.R
✔ |   5       | tests/test_collapse_rp_lists.R

══ Results ═════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 3.5 s

OK:       51
Failed:   0
Warnings: 0
Skipped:  0
Done running run-tests.sh
```

logstar commented:
@NHJohnson The error is caused by readr update, <https://www.tidyverse.org/blog/2018/12/readr-1-3-1/#tibble-subclass>, so the read-in table is of class `spec_tbl_df`. Will update for compatibility.

NHJohnson commented:
> @NHJohnson The error is caused by readr update, https://www.tidyverse.org/blog/2018/12/readr-1-3-1/#tibble-subclass, so the read-in table is of class `spec_tbl_df`. Will update for compatibility.

Ok thanks! Can you point me to the Docker image you used to run the tests?

logstar commented:
@NHJohnson There is also an error in the failed tests, which is also caused by the change of `testthat::expect_equal` behavior. Docker image/container version does not check column order, whereas the latest version checks. Will fix the tests soon.

logstar commented:
Fixed test compatibility issues. All test should past with recent package versions. Following is a run on my host machine, which has recent R and packages versions.

```text
$ bash annotator/run-tests.sh 
✔ |  OK F W S | Context
✔ |  31       | tests/test_annotate_long_format_table.R [3.6 s]                                                                                                                         
✔ |   8       | tests/test_collapse_name_vec.R                                                                                                                                          
✔ |   7       | tests/test_collapse_rp_lists.R                                                                                                                                          
✔ |   5       | tests/test_collapse_rp_lists.R                                                                                                                                          

══ Results ═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 3.8 s

[ FAIL 0 | WARN 0 | SKIP 0 | PASS 51 ]
Done running run-tests.sh
```

NHJohnson reviewed (approved):
Ok thank you, all the tests now pass for me.

logstar commented:
> Ok thank you, all the tests now pass for me.

Thank you again for reviewing and test-running @NHJohnson !
